floodscan_validation/exploration/threshold_deciding.R

```diff
@@ -0,0 +1,1083 @@
+  ) %>%
+  mutate(
+    rp = as.factor(rp),
+    value = replace_na(value, 0),
+    best_rp = value == max(value, na.rm = TRUE),
+    basin_name = factor(
+      case_when(
+        basin_id == 1040760290 ~ "Upper Niger",
+        basin_id == 1040909890 ~ "Lower Niger",
+        basin_id == 1040022420 ~ "Niger Delta",
+        basin_id == 1040909900 ~ "Benue"
+      ),
+      levels = c("Upper Niger", "Lower Niger", "Benue", "Niger Delta")
+    )
+  ) %>%
+  group_by(
+    basin_id,
+    name,
+  ) %>%
+  mutate(
+    best_overall = value == max(value, na.rm = TRUE)
+  )
+
+
+base_metrics_plot <- df_metrics_plot %>%
+  ggplot(
+    aes(
+      x = threshold,
+      y = rp,
+      fill = value
+    )
+  ) +
+  geom_tile(
+    color = "white"
+  ) +
+  facet_grid(
+    rows = vars(basin_name),
+    cols = vars(name),
+    labeller = labeller(.cols = str_to_title)
+  ) +
+  scale_x_continuous(
+    labels = scales::label_percent()
+  ) +
+  scale_fill_gradient(
+    low = "white",
+    high = hdx_hex("mint-hdx")
+  ) + 
+  theme(
+    panel.grid = element_blank()
+  ) +
+  labs(
+    x = "Threshold",
+    y = "Return period",
+    fill = "Metric"
+  )
+
+base_metrics_plot
+
+# look at overall best perofmring options
+base_metrics_plot + geom_tile(
+  data = filter(df_metrics_plot, best_overall),
+  fill = NA,
+  color = hdx_hex("gray-dark"),
+  lwd = .5
+)
+
+# look at best performing by RP if we want to look at RPs
+
+base_metrics_plot + geom_tile(
+  data = filter(df_metrics_plot, best_rp),
+  fill = NA,
+  color = hdx_hex("gray-dark"),
+  lwd = .5
+)
+
+
+# look at overall metrics across all basins
+
+df_metrics_overall <- df_metrics_plot %>%
+  group_by(
+    threshold,
+    rp
+  ) %>%
+  summarize(
+    across(
+      .cols = tp:fn,
+      .fns = sum
+    )
+  ) %>%
+  mutate(
+    precision = tp / (tp + fp),
+    recall = tp / (tp + fn),
+    f1 = precision * recall / (precision + recall)
+  ) %>%
+  pivot_longer(
+    precision:f1
+  ) %>%
+  group_by(
+    rp,
+    name
+  ) %>%
+  mutate(
+    best = value == max(value)
+  )
+
+df_metrics_overall %>%
+  ggplot(
+    aes(
+      x = threshold,
+      y = rp,
+      fill = value
+    )
+  ) +
+  geom_tile() +
+  facet_wrap(
+    ~ name,
+    labeller = labeller(.cols = str_to_title)
+  ) +
+  scale_x_continuous(
+    labels = scales::label_percent()
+  ) +
+  scale_fill_gradient(
+    low = "white",
+    high = hdx_hex("mint-hdx")
+  ) + 
+  scale_color_manual(
+    values = c("white", "black")
+  ) +
+  theme(
+    panel.grid = element_blank()
+  ) +
+  labs(
+    x = "Threshold",
+    y = "Return period",
+    fill = "Metric"
+  )
+
+#################################################################
+#### CHECK FREQUENCY OF ACTIVATION BASED ON THESE THRESHOLDS ####
+#################################################################
+
+final_activations <- basin_pcts %>%
+  mutate(
+    activation_2y = `2y_rp` >= 1,
+    activation_5y = `2y_rp` >= 0.5,
+    activation_20y = `2y_rp` > 0.2,
```

Comment:
is this correct ? shouldn't it be

```
activation_2y = `2yr_rp >=0.8
activation_5y = `5yr_rp >=0.5
activation_2y = `20yr_rp >=0.2
```

Also from the plots above the 80 % threshold decision was not super clear. 

I made a plot where I plotted x-axis thresholds, then TP,FP,FN (left y-axis) and f1 + precision + recall on right axis. The optimal threshold decision became a bit less clear even.

Reply:
lol to the first thing, yes, you are correct, I was playing around to see how different thresholds behave and didn't mean to actually save that.

Reply:
Nice plot! I think the reality behind why we chose 80% is exactly what you showed there, which is that performance is actually fairly consistent across most values. Overall does tend to be better on balance (F1) as the threshold increases, but shows decreasing performance in some basins when you require 100% thresholds. But the reality behind all of it is we can definitely not decrease our thresholds because they just activate much too frequently, and we don't want to increase because even though in some basins like Benue it increases the performance, across the basins as a whole we have optimal performance around 80%.